pkg/bayesiantrain/trainer.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/crowdsecurity/crowdsec/pkg/types"
+)
+
+var exprCache map[string]vm.Program
+var exprCacheLock sync.Mutex
+
+type LogEventStorage struct {
+	ParsedIpEvents map[string]fakeBucket
+	labeled        bool
+	total          int
+	nEvilIps       int
+	nBenignIps     int
+}
+
+type fakeBucket struct {
+	events []types.Event
+	leaky  *leakybucket.Leaky
+	label  int
+}
+
+type evalIpResult struct {
+	Result int
+	Label  int
+	Error  error
+}
+
+type evalHypothesisResult struct {
+	EvilIpsWithHypothesis   int
+	BenignIpsWithHypothesis int
+	Error                   error
+}
+
+func evaluateProgramOnBucket(l *fakeBucket, compiledExpr *vm.Program, outputChan chan<- evalIpResult) {
+	var hypothesis, ok bool
+	for index, evt := range l.events {
+		ret, err := expr.Run(compiledExpr, map[string]interface{}{"evt": &evt, "queue": leakybucket.Queue{Queue: l.events[:index], L: index + 1}, "leaky": &(l.leaky)})
+		if err != nil {
+			outputChan <- evalIpResult{Error: err}
+		}
+
+		if hypothesis, ok = ret.(bool); !ok {
+			outputChan <- evalIpResult{Error: fmt.Errorf("bayesion hypothesis, unexpected non-bool return : %T", ret)}
+		}
+
+		if hypothesis {
+			outputChan <- evalIpResult{Result: 1, Label: l.label}
+			break
+		}
+	}
+
+	outputChan <- evalIpResult{Result: 0, Label: l.label}
+}
+
+func controllerRoutine(inputChan <-chan evalIpResult, outputChan chan<- evalHypothesisResult, totalIps int) {
+
+	var evilIpsWithHypothesis = 0
+	var benignIpsWithHypothesis = 0
+	var totalIpsSeen = 0
+	for {
+		result := <-inputChan
+
+		if result.Error != nil {
+			outputChan <- evalHypothesisResult{Error: result.Error}
+			break
+		}
+
+		if result.Label == 0 {
+			benignIpsWithHypothesis += result.Result
+		}
+		if result.Label == 1 {
+			evilIpsWithHypothesis += result.Result
+		}
+		totalIpsSeen += 1
+
+		if totalIpsSeen == totalIps {
+			break
+		}
+	}
+
+	outputChan <- evalHypothesisResult{EvilIpsWithHypothesis: evilIpsWithHypothesis, BenignIpsWithHypothesis: benignIpsWithHypothesis, Error: nil}
+}
+
+func (s *LogEventStorage) InsertLabels(evilIps []string) {
+
+	var nUsedLabels = 0
+	for _, ip := range evilIps {
+		if parsedEvent, ok := s.ParsedIpEvents[ip]; ok {
+			parsedEvent.label = 1
+			s.ParsedIpEvents[ip] = parsedEvent
+			nUsedLabels += 1
+		}
+	}
+	s.nEvilIps = nUsedLabels
+	s.nBenignIps = s.total - s.nEvilIps
+	s.labeled = true
+}
+
+func (s *LogEventStorage) TestHypothesis(hypothesis string) error {
+
+	if !s.labeled {
+		return fmt.Errorf("LogEventStorage has not been labeled yet, add labels using InsertLabels")
+	}
+	compiled, err := compileAndCacheHypothesis(hypothesis)
+	if err != nil {
+		return err
+	}
+
+	inputChan := make(chan evalIpResult, 1000)
+	outputChan := make(chan evalHypothesisResult)
+	defer close(inputChan)
+	defer close(outputChan)
+
+	for _, v := range s.ParsedIpEvents {
+		go evaluateProgramOnBucket(&v, compiled, inputChan)
+	}
+
+	go controllerRoutine(inputChan, outputChan, s.total)
```

Comment:
I’m not aware of the larger context of where this fits in. Looks like you’re adding new function in scenarios/parsers. That being said  I think using go routines won't add performance but is only making the code more complex. The “go-routined” function `evaluateProgramOnBucket` is not doing any I/O work hence doesn't make sense to use go routines on it. Something like
```go
var result evalHypothesisResult
for _, v := range s.ParsedIpEvents {
 r := evaluateProgramOnBucket(&v, compiled, inputChan)
// update result
}
```    
Would have equivalent if not better performance while eliminating the overhead of controller routine etc

Reply:
Ok maybe I should explain. The whole thing is the main training loop for the bayesian buckets. The construction works the following way:
* The logs are loaded into the `LogEventStorage` into the map with the key being the Ip and the all the events for this Ip are added to the `fakeBucket`
* The user can then test different hypothesis expr to see if any of them would make for good conditions in the bucket using `TestHypothesis`
* To speed up the hypothesis testing the idea way to run it in parallel threads for each IP (as its basically counting some stuff per IP) 
* The goal of the channel/goroutine design in `TestHypothesis` is to enable this parallelism by spawing an independent routine for each IP (fake bucket) and then collecting all the results using the controller.  

Does this make more sense now?

Reply:
Understood, the goroutines would definitely increase throughput. If the training is CPU intensive task than goroutine would make sense.

Reply:
Nice, thank you